### PR TITLE
Add e2e test case to restore trashed post

### PR DESCRIPTION
**Pull request by pooja-muchandikar**
Trac ticket: https://core.trac.wordpress.org/ticket/54843

### CODE REVIEW

**Comment by JustinyAhin** [comment]
Hello @pooja-muchandikar,

Here are a few things that I noticed and that could be improved in your PR:
- Test files end with `.test.js`, so your test file would become `restore-trash-post.test.js`
- `loginUser()` is not used in the tests, so it can be removed from the imports
- It is a good idea to remove all posts before your tests (see https://github.com/WordPress/wordpress-develop/blob/371966a187081662c155ef98ddb7a681dc10f268/tests/e2e/specs/edit-posts.test.js#L10-L12 for example)

**Comment by JustinyAhin** [comment]
@pooja-muchandikar thanks for adding the updates. Your tests are very clean and they all pass. Here is something that we could improve.

- Here https://github.com/pooja-muchandikar/wordpress-develop/blob/c8e28fc74f1d074f46caa7849796490b1743cd2a/tests/e2e/specs/empty-trash-posts.test.js#L40: 

It might make sense to assert the message that appears when the post is trashed. In this case, it'll be "1 post permanently deleted."
That will help in case the trashing steps outputs an error. By asserting the success message, we make sure that we have the expected behavior.

**Comment by JustinyAhin** [comment]
Except the above comment, the PR looks fine to me.

@tellthemachines @hellofromtonya you might want to give a look, before we move to merging it.

**Comment by pooja-muchandikar** [comment]
Hi @JustinyAhin,

Previously, assertion message was only added, but while running the test cases CI was failing, as the number of posts were getting deleting dynamically, that's the reason changed the selector. 

https://github.com/WordPress/wordpress-develop/pull/2186/commits/3b817af66d229507042706bb9125274a5d07823f#diff-3860c93f38e576535aad8fdbb45ee1be20e197c829f616f26609cfb253910d6eL47

**Comment by pooja-muchandikar** [comment]
Hi @JustinyAhin, @hellofromtonya

I hope you are doing well!

Wanted to know if there is any update on this? Do I need to make the changes or can we proceed further?
> https://github.com/WordPress/wordpress-develop/pull/2186#issuecomment-1041162514

**Comment by JustinyAhin** [comment]
> @tellthemachines @hellofromtonya you might want to give a look, before we move to merging it.

Hello @pooja-muchandikar, the PR is fine for me. Can you bring it up during the next Test team meeting on Thursday, so it can be reviewed by a committer?

**Comment by pooja-muchandikar** [comment]
Sure
Thanks!

File: tests/e2e/specs/restore-trash-post.test.js
Comment (kevin940726) [inline tests/e2e/specs/restore-trash-post.test.js]: Isn't this the same in the `empty-trash-posts` test spec?

I wonder why we would want to have different test specs for them? I guess it's okay to write them in the same file?

File: tests/e2e/specs/empty-trash-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-posts.test.js]: Nit: Styling.

```suggestion

        // Create a Post
```

File: tests/e2e/specs/empty-trash-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-posts.test.js]: Might be better if we specifically target the text of the tab using xPath: `'//a[text()="Trash"]'`.

File: tests/e2e/specs/empty-trash-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-posts.test.js]: I don't think we need to wait for this element?

File: tests/e2e/specs/empty-trash-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-posts.test.js]: Same here, might be better to target the text of the element.

```js
const deleteAllButton = await page.waitForSelector('input[value="Empty Trash"]');
await deleteAllButton.click();
```

File: tests/e2e/specs/empty-trash-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-posts.test.js]: We should add an assertion for the message to make the test easier to read and have a clear expectation of the result. Any reason why we can't add an assertion here?

File: tests/e2e/specs/restore-trash-post.test.js
Comment (kevin940726) [inline tests/e2e/specs/restore-trash-post.test.js]: Nit: Missing an indention.

```suggestion
		await page.click("a[href='edit.php?post_status=trash&post_type=post']");
```

File: tests/e2e/specs/empty-trash-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-posts.test.js]: I don't think we have to wait for this element?

File: tests/e2e/specs/restore-trash-post.test.js
Comment (kevin940726) [inline tests/e2e/specs/restore-trash-post.test.js]: Let's use `aria-label` to target the element:

```js
await page.click(`[aria-label="Restore “${title}” from the Trash"]`);
```

File: tests/e2e/specs/restore-trash-post.test.js
Comment (kevin940726) [inline tests/e2e/specs/restore-trash-post.test.js]: We can just target the whole `#message` and use `toContain` or `toMatch` to assert the match.

```js
const messageElement = await page.waitForSelector('#message');

expect(
    await messageElement.evaluate( ( element ) => element.innerText )
).toContain( '1 post restored from the Trash.' )
```

File: tests/e2e/specs/empty-trash-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-posts.test.js]: I don't think we should do this in `beforeEach` though, I think it's better to do it in `afterEach`. Tests should make sure to clear up all states after they pass anyway.

Ideally, this should call the REST API, but I understand it's not currently available in `e2e-tests-utils`.

**Review by kevin940726 — COMMENTED** [review]
Thank you for working on this, and your patience for the late review!

Left some comments, but note that many of them are only suggestions, and the experience of writing a more readable test should be improved once we [migrate to Playwright](https://github.com/WordPress/gutenberg/pull/38570). In the meantime, this is a great start!

File: tests/e2e/specs/empty-trash-posts.test.js
Comment (pooja-muchandikar) [reply tests/e2e/specs/empty-trash-posts.test.js]: Yes I have specified the reason [here](https://github.com/WordPress/wordpress-develop/pull/2186#issuecomment-1041162514)

Comment (kevin940726) [reply tests/e2e/specs/empty-trash-posts.test.js]: > the number of posts were getting deleting dynamically.

I think I'm missing something here, but why would the number of posts be dynamic? Is it because we're not clearing up states between tests? Or are we parallelizing tests that causes race-condition?

**Comment by pooja-muchandikar** [comment]
Hi @kevin940726,

I have addressed all the feedbacks shared by you and also combined the test cases, please take a look. 

Thanks!

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: Ideally, we should still make an assertion about the confirmation message here.

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: I think we missed [this](https://github.com/WordPress/wordpress-develop/pull/2186#discussion_r815929153) 😅 .

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: Same here.

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (kevin940726) [inline tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: I think we missed [this](https://github.com/WordPress/wordpress-develop/pull/2186#discussion_r815960434) 😅 .

**Review by kevin940726 — COMMENTED** [review]
Overall it looks good!

We want to pay attention to the [styling guide](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/) though, like using single quotes and remove redundant whitespaces/line-breaks. We should definitely enable linting at some point though so I'm not gonna nitpick them in this PR 🤣.

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (pooja-muchandikar) [reply tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: Hi @kevin940726,

This feedback is addressed

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (pooja-muchandikar) [reply tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: This feedback is addressed, some we tried to use xpath as suggested by you but unfortunately using xpath selector is leading to failing CI, in my local it is working fine but in CI its failing.

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (pooja-muchandikar) [reply tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: This feedback is addressed, some we tried to use xpath as suggested by you but unfortunately using xpath selector is leading to failing CI, in my local it is working fine but in CI its failing.

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (pooja-muchandikar) [reply tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: Hi @kevin940726,

I tried to make an assertion about confirmation message but unfortunately it didn't worked, when I am adding a assertion message the number of deleted post are dynamic(sometime value is 7 sometime 8,9) so I had to use this. :(

**Comment by pooja-muchandikar** [comment]
Hi @kevin940726,

Made all necessary changes and also commented on respective feedbacks with reasoning. Also tried to work on few indention changes please take a look. 

Thanks!

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (kevin940726) [reply tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: I think it's because we aren't resetting the state after each test correctly. `trashAllPosts` only puts all active posts into the trash, but doesn't delete them.

We'll need a `deleteAllPosts` util similar to what I've done in
[Playwright](https://github.com/WordPress/gutenberg/pull/38570/files#diff-b0be83d71db6ce709700ada10e1f3cef99d8d59b2c30c11861a96e083e68dc7d) to force delete all posts after each test. For now, I think we can use the `__experimentalRest` api to create one in this file, and replace that with `deleteAllPosts` once we migrate to playwright.

If that's too much work for now, I'm okay if we just add a `// TODO` comment above and leave that to our future selves.

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (kevin940726) [reply tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: Why is it failing in CI? Do you have the error messages? We should probably fix that instead of doing it the less preferable way.

Comment (pooja-muchandikar) [reply tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: @kevin940726, As per your suggestion we used xpath, and the xpath did pass in local but on pushing the code, it failed on CI.
Screenshot 2022-03-03 at 3 23 14 PM

File: tests/e2e/specs/empty-trash-restore-trashed-posts.test.js
Comment (pooja-muchandikar) [reply tests/e2e/specs/empty-trash-restore-trashed-posts.test.js]: Okay @kevin940726, I have added `// TODO` for now.

**Comment by pooja-muchandikar** [comment]
Hi @kevin940726,

I hope you are doing well! Just checking if you checked latest commits on this PR.

**Review by kevin940726 — APPROVED** [review]
Looks great! Thanks again for the hard work!

**Comment by hellofromtonya** [comment]
Committed via changeset  https://core.trac.wordpress.org/changeset/52839. Thank you @pooja-muchandikar @JustinyAhin @kevin940726 for your contributions ⭐

**Pull request closed without merging**
